design/reboot-interface.md

```diff
@@ -0,0 +1,193 @@
+sequential changes to the spec).
+
+## Motivation
+
+We require a non-destructive way to fence a Kubernetes node. Some nodes cannot
+be replaced (it is currently not possible for a new master to join a cluster),
+or are expensive to replace (e.g. if this would require rebalancing Ceph data).
+A solution to this is for fencing to reboot the Host, thus ensuring that
+running processes are stopped to avoid a split-brain scenario while still
+allowing the node to rejoin the cluster with its data intact (albeit stale)
+after the reboot.
+
+The expected implementation of this on the fencing side is a Custom Resource
+Definition that requests a Machine be rebooted. The actual reboot will be
+effected either by the Machine Actuator itself (if the Cluster-API SIG can be
+persuaded that this should be part of the API in the long term), or some
+equivalent of the Machine Actuator limited to just this purpose.
+
+### Goals
+
+The Machine (Reboot) Actuator will require from the BareMetalHost:
+
+* A declarative API to perform the reboot
+* The ability to determine a point where all processes running on the Host at
+  the time of the fencing decision are guaranteed to have been stopped
+* Rapid convergence to a state where all running processes are stopped,
+  independent of other happenings in the system
+
+### Non-Goals
+
+There is no requirement to implement a scheduled reboot. In the Kubernetes
+context, reboot decisions should generally be made in realtime by some higher
+level of the Node/Machine hierarchy, to take into account such questions as the
+overall health of the cluster and the effect of a reboot on that. The best
+implementation for this would be a RebootSchedule CRD that waits until the
+appointed time before issuing an immediate reboot request to the BareMetalHost.
+This allows multiple reboots to be scheduled, scheduled reboots to be
+manipulated or cancelled, and a record to be left behind of past scheduled
+reboots. The proposed design could easily be extended to accomodate this
+requirement should it arise in future.
+
+There is no requirement to hold the power off for any particular length of
+time. For fencing purposes we only need to be sure when the node is powered
+off.
+
+This API is not responsible for managing unprovisioned hosts, e.g. to recover
+from state inconsistencies in Ironic prior to provisioning. Any such
+inconsistencies that are not handled internally by the BareMetalHost will
+likely need manual intervention to recover anyway.
+
+## Proposal
+
+A new date-time field, ``lastPoweredOn``, will be added to the ``provisioning``
+section of the BareMetalHost status. This records a time after which the Host
+was last booted using the current image. Processes running prior to this time
+may be assumed to have been stopped.
+
+To provide a declarative interface for performing a reboot, a new
+BareMetalHostRebootRequest CRD will be created. The spec of this custom
```

Comment:
What do you think about just calling this `BareMetalHostReboot`? Since k8s has a declarative API, every resource is merely a "request".

A more common example of an imperative action is "backup". Operators such as etc-operator tend to have resources called `EtcdBackup` or similar.

Reply:
I was going to ask you about the name, since it's so long... do we need the BareMetal part either, given that the resource will already be under `metal3.io/`?
I'm not a fan of naming declarative APIs with verbs, but I'd go along with it if that's the pattern. 'Backup' can either be a verb or a noun though, and I suspect it's being used as a noun in that example (come to think of it, "back up" as a verb is two words - you can "back up some data" or "make a backup of some data" - so I'd expect to see `BackUpEtcd` if it were being used as a verb).

Reply:
> I was going to ask you about the name, since it's so long... do we need the BareMetal part either, given that the resource will already be under `metal3.io/`?

This is an area of ambiguity. You would think the namespacing would be enough. But the group and version can be omitted when using a client like `kubectl`, and that can introduce ambiguity. What if some other operator, like MCO, also create a `Reboot` resource, and then I do `kubectl get reboots`? kubectl seems to pick one of them somewhat arbitrarily.

In practice, many operators address this with a prefix on their resource, such as `EtcdBackup` does. I suggest we do that here.

> I'm not a fan of naming declarative APIs with verbs, but I'd go along with it if that's the pattern. 'Backup' can either be a verb or a noun though, and I suspect it's being used as a noun in that example (come to think of it, "back up" as a verb is two words - you can "back up some data" or "make a backup of some data" - so I'd expect to see `BackUpEtcd` if it were being used as a verb).

Reboot can also [be a noun](https://www.merriam-webster.com/dictionary/reboot).

Reply:
True! But for some reason you would never hear anybody say "let's create a reboot" or "let's delete that reboot" in the way that you might "create an etcd backup" or "delete an etcd backup".
Do you think `HostRebootRequest` would be sufficiently unique?
I guess we always have the option of specifying multiple short names that we can use in practice even if the official one is painfully long.

Reply:
> True! But for some reason you would never hear anybody say "let's create a reboot" or "let's delete that reboot" in the way that you might "create an etcd backup" or "delete an etcd backup".

It's normal to talk about creating a backup because the act of performing a backup produces something. Performing a reboot does not normally produce anything.

But it's perfectly normal to ask someone to perform a reboot.

As another example, here's an interesting feature of journalctl:

```
$ journalctl --list-boots
 -2 04daa48a0ba249089166eed69fa913b8 Wed 2019-07-31 23:03:59 EDT—Fri 2019-08-02 18:23:19 EDT
 -1 25660fb316124c0ca6de986ee0fc76ba Fri 2019-08-02 18:23:37 EDT—Tue 2019-08-06 13:54:33 EDT
  0 ece6a74488224946a59960a2675d4aed Tue 2019-08-06 13:54:54 EDT—Tue 2019-08-06 15:59:16 EDT
```

Nobody created or deleted those "boots", which may have been "reboots", but they sure happened. We can list them and see their properties.

> Do you think HostRebootRequest would be sufficiently unique?

I would go with BareMetalHostReboot, and likely include a short name of `bmhreboot`. HostReboot would probably be unique enough to avoid direct collision, but it could still cause confusion if there are any other "host"-related APIs from some other api group.